### PR TITLE
ZEPPELIN-3681. Introduce NotebookService

### PR DESCRIPTION
**Pull request by zjffdu**
### What is this PR for?
This the first phase of refactoring work of ZEPPELIN-3288. The background of ZEPPELIN-3288 is that zeppelin provides 2 kinds of api (rest api and websocket), both of them invoke the same backend logic.  This PR put all the notebook related operation into class NotebookService which called by both NotebookServer & NotebookRestApi.

### What type of PR is it?
[Refactoring]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-3288

### How should this be tested?
* CI pass

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by zjffdu** [comment]
@jongyoul @felixcheung Could you help review it ? Thanks

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 1]
```diff
@@ -894,35 +882,21 @@ private void updateNote(NotebookSocket conn, HashSet<String> userAndRoles, Noteb
...
-        notebook.refreshCron(note.getId());
-      }
-
-      AuthenticationInfo subject = new AuthenticationInfo(fromMessage.principal);
-      note.persist(subject);
-      broadcast(note.getId(), new Message(OP.NOTE_UPDATED).put("name", name).put("config", config)
-          .put("info", note.getInfo()));
-      broadcastNoteList(subject, userAndRoles);
-    }
+    NotebookService notebookService = new NotebookService(notebook());
```
Comment (jongyoul) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: Why doesn't this method use `getNotebookService()`?

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 2]
```diff
@@ -1157,30 +1093,20 @@ private void moveFolderToTrash(NotebookSocket conn, HashSet<String> userAndRoles
...
-  private void restoreNote(NotebookSocket conn, HashSet<String> userAndRoles, Notebook notebook,
-          Message fromMessage) throws SchedulerException, IOException {
+  private void restoreNote(NotebookSocket conn,
+                           Message fromMessage) throws IOException {
     String noteId = (String) fromMessage.get("id");
-
     if (noteId == null) {
       return;
     }
+    NotebookService notebookService = new NotebookService(notebook());
```
Comment (jongyoul) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: this, too.

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 3]
```diff
@@ -1222,57 +1148,42 @@ private void restoreAll(NotebookSocket conn, HashSet<String> userAndRoles, Noteb
...
       return;
     }
-
-    Map<String, Object> params = (Map<String, Object>) fromMessage.get("params");
-    Map<String, Object> config = (Map<String, Object>) fromMessage.get("config");
     String noteId = getOpenNoteId(conn);
     if (noteId == null) {
       noteId = (String) fromMessage.get("noteId");
     }
+    NotebookService notebookService = new NotebookService(notebook());
```
Comment (jongyoul) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: this, too.

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 4]
```diff
@@ -2601,53 +2461,41 @@ public void clearParagraphRuntimeInfo(InterpreterSetting setting) {
...
-
-    Note note = notebook.getNote(noteId);
-    if (note != null) {
-      note.setNoteParams(noteParams);
-
-      AuthenticationInfo subject = new AuthenticationInfo(fromMessage.principal);
-      note.persist(subject);
-      broadcastNoteForms(note);
-    }
+    NotebookService notebookService = new NotebookService(notebook());
```
Comment (jongyoul) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: here, too.

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 5]
```diff
@@ -2601,53 +2461,41 @@ public void clearParagraphRuntimeInfo(InterpreterSetting setting) {
...
-    Note note = notebook.getNote(noteId);
-    if (note != null) {
-      note.getNoteForms().remove(formName);
-      note.getNoteParams().remove(formName);
-
-      AuthenticationInfo subject = new AuthenticationInfo(fromMessage.principal);
-      note.persist(subject);
-      broadcastNoteForms(note);
-    }
+    NotebookService notebookService = new NotebookService(notebook());
```
Comment (jongyoul) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: here, too.

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 6]
```diff
@@ -1327,142 +1238,141 @@ private void patchParagraph(NotebookSocket conn, HashSet<String> userAndRoles,
...
+          }
+        });
+  }
 
-    if (!hasParagraphWriterPermission(conn, notebook, noteId,
-        userAndRoles, fromMessage.principal, "clear output")) {
-      return;
-    }
+  protected Note importNote(NotebookSocket conn, Message fromMessage) throws IOException {
+    NotebookService notebookService = new NotebookService(notebook());
```
Comment (jongyoul) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: here, too.

**Comment by jongyoul** [comment]
Basically, I agree with the structure you imagine. This might have some style issues, but I don't think it's not a blocker for this stage. LGTM

**Comment by zjffdu** [comment]
Thanks @jongyoul I have addressed the comments

**Pull request closed without merging**